Author: Michael Webster <miketwebster@gmail.com>
Date: Wed, 15 Nov 2017 20:39:57 -0500
Description: dnd: Make sure to clear any source_fs pointers after dnd operations.

 The drag info struct itself lives as long as the container (I made an
 incorrect assumption previously that it gets cleared after dnd ops.)

 This was causing an issue in nemo-dnd.c when getting the default drag
 action by comparing filesystem ids (the original was never cleared for
 subsequent operations, so the first action type was always re-used, even
 if the source filesystem really did change.) - line 502.

 Fixes #1642

Origin: upstream, https://github.com/linuxmint/nemo/commit/80ca131a2c14039a7599f6e7f87b7469225de9e2
Last-Update: 2017-11-16
---
 libnemo-private/nemo-dnd.c      | 4 ----
 libnemo-private/nemo-icon-dnd.c | 3 +++
 src/nemo-window-slot-dnd.c      | 1 +
 3 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/libnemo-private/nemo-dnd.c b/libnemo-private/nemo-dnd.c
index 8be9bff35..588afda51 100644
--- a/libnemo-private/nemo-dnd.c
+++ b/libnemo-private/nemo-dnd.c
@@ -395,10 +395,6 @@ check_same_fs (NemoFile    *target_file,
 
     result = FALSE;
 
-    if (target_file != NULL && source_fs_for_desktop != NULL) {
-
-    }
-
     if (target_file != NULL && source_file != NULL) {
         source_id = nemo_file_get_filesystem_id (source_file);
         target_id = nemo_file_get_filesystem_id (target_file);
diff --git a/libnemo-private/nemo-icon-dnd.c b/libnemo-private/nemo-icon-dnd.c
index 7e5600f91..47fda7e2b 100644
--- a/libnemo-private/nemo-icon-dnd.c
+++ b/libnemo-private/nemo-icon-dnd.c
@@ -581,6 +581,7 @@ drag_end_callback (GtkWidget *widget,
 
 	nemo_drag_destroy_selection_list (dnd_info->drag_info.selection_list);
 	dnd_info->drag_info.selection_list = NULL;
+    g_clear_pointer (&dnd_info->drag_info.source_fs, g_free);
 }
 
 static NemoIcon *
@@ -1415,6 +1416,8 @@ nemo_icon_container_receive_dropped_icons (NemoIconContainer *container,
 
 	g_free (drop_target);
 	nemo_drag_destroy_selection_list (container->details->dnd_info->drag_info.selection_list);
+
+    g_clear_pointer (&container->details->dnd_info->drag_info.source_fs, g_free);
 	container->details->dnd_info->drag_info.selection_list = NULL;
     free_dnd_grid (container);
 }
diff --git a/src/nemo-window-slot-dnd.c b/src/nemo-window-slot-dnd.c
index 559d1ac10..442171a2c 100644
--- a/src/nemo-window-slot-dnd.c
+++ b/src/nemo-window-slot-dnd.c
@@ -149,6 +149,7 @@ drag_info_clear (NemoDragSlotProxyInfo *drag_info)
 
   if (drag_info->info == NEMO_ICON_DND_GNOME_ICON_LIST) {
     nemo_drag_destroy_selection_list (drag_info->data.selection_list);
+    g_clear_pointer (&drag_info->desktop_dnd_source_fs, g_free);
   } else if (drag_info->info == NEMO_ICON_DND_URI_LIST) {
     g_list_free (drag_info->data.uri_list);
   } else if (drag_info->info == NEMO_ICON_DND_NETSCAPE_URL) {
